python/tvm/relay/op/_transform.py

```diff
@@ -56,3 +59,105 @@
 # layout_transform
 _reg.register_schedule("layout_transform", schedule_injective)
 _reg.register_pattern("layout_transform", OpPattern.INJECTIVE)
+
+# shape func
+@script
+def _arange_shape_func(start, stop, step):
+    out = output_tensor((1,), "int64")
+    out[0] = int64(ceil_div((float32(stop[0]) - float32(start[0])), float32(step[0])))
+    return out
+
+@_reg.register_shape_func("arange", True)
+def arange_shape_func(attrs, inputs, _):
+    return [_arange_shape_func(*inputs)]
+
+@script
+def _concatenate_shape_func(inputs, axis):
+    ndim = inputs[0].shape[0]
+    out = output_tensor((ndim,), "int64")
+    for i in const_range(ndim):
+        if i != axis:
+            out[i] = inputs[0][i]
+            for j in const_range(1, len(inputs)):
+                assert out[i] == inputs[j][i], \
+                    "Dims mismatch in the inputs of concatenate."
+        else:
+            out[i] = int64(0)
+            for j in const_range(len(inputs)):
+                out[i] += inputs[j][i]
+    return out
+
+@_reg.register_shape_func("concatenate", False)
+def concatenate_shape_func(attrs, inputs, _):
+    axis = get_const_int(attrs.axis)
+    return [_concatenate_shape_func(inputs, convert(axis))]
+
+@script
+def _reshape_shape_func(x, newshape, ndim):
+    out = output_tensor((ndim,), "int64")
+    src_idx = 0
+    dst_idx = 0
+    infer_idx = -1
+    copy = False
+    skip = 0
+    for i in const_range(len(newshape)):
+        if skip > 0:
+            skip -= 1
+        elif newshape[i] > 0:
+            out[dst_idx] = int64(newshape[i])
+            src_idx += 1
+            dst_idx += 1
+        elif newshape[i] == 0:
+            out[dst_idx] = x[src_idx]
+            src_idx += 1
+            dst_idx += 1
+        elif newshape[i] == -1:
+            assert infer_idx < 0, "One and only one dim can be inferred"
+            out[dst_idx] = int64(1)
+            infer_idx = i
+            dst_idx += 1
+        elif newshape[i] == -2:
+            copy = True
+        elif newshape[i] == -3:
+            assert x.shape[0] - src_idx > 1, \
+                "Not enough dims in input shape for -3"
+            out[dst_idx] = x[src_idx] * x[src_idx+1]
+            src_idx += 2
+            dst_idx += 1
```

Comment:
Are those only for the very ad-hoc reshape functions in MXNet?

Reply:
Yes

Reply:
This might be out of the scope of this PR but can we think about removing the ad-hoc reshape from the Relay core and possibly elaborating away i.e MxNet dialect.